### PR TITLE
Dropdowns v2

### PR DESCRIPTION
**Pull request by actioninja**
## About The Pull Request

Complete rewrite of dropdowns to use popper elements that are injected at the body level.
This means they are now horribly overcomplicated and also suck a lot less

Also ported `Icon` to typescript for a reason that didn't end up making it in the PR

## Why It's Good For The Game
image
image

## Changelog
:cl:
add: good tgui dropdowns
del: tgui dropdown suck
qol: tgui dropdowns suck less
balance: buffed tgui dropdowns
/:cl:

### CODE REVIEW

**Comment by Mothblocks** [comment]
why are you giving yourself good first pr

**Comment by LemonInTheDark** [comment]
image
am I stupid or is the padding here different. Is that a preferences menu thing or part of this pr

**Comment by actioninja** [comment]
> image am I stupid or is the padding here different. Is that a preferences menu thing or part of this pr

The padding breaks randomly, it did this before, I have no idea why, blame IE.

**Review by stylemistake — APPROVED** [review]
Nice

File: tgui/packages/tgui/components/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  width: 0,
+  height: 0,
+  top: 0,
+  right: 0,
+  bottom: 0,
+  left: 0,
+  x: 0,
+  y: 0,
+  toJSON: () => null,
+};
```
Comment (jlsnow301) [inline tgui/packages/tgui/components/Dropdown.tsx:50]:
```suggestion
} as const;
```
Good idea unless it gives type errors elsewhere

File: tgui/packages/tgui/components/Dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+      Dropdown.renderedMenu = renderedMenu;
+    }
+
+    const domNode = this.getDOMNode()!;
+    Dropdown.currentOpenMenu = domNode;
+
+    renderedMenu.scrollTop = 0;
+    renderedMenu.style.width =
+      // Hack, but domNode should *always* be the parent control meaning it will have width
+      // @ts-ignore
```
Comment (jlsnow301) [inline tgui/packages/tgui/components/Dropdown.tsx:111]: Really wish ts did not get disabled, despite this- is it possible to use an `in` here? `"menuWidth" in this.props`

File: tgui/packages/tgui/interfaces/NavBeacon.tsx [hunk 3]
```diff
@@ -25,7 +25,7 @@ export type DisabledProps = {
 };
 
 export type NavBeaconStaticControl = {
-  direction_options: String[];
+  direction_options: string[];
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/NavBeacon.tsx:28]: Nice find

**Review by jlsnow301 — COMMENTED** [review]
Nothing major, looks okay

File: tgui/packages/tgui/components/Dropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,297 @@
+import { createPopper, VirtualElement } from '@popperjs/core';
+import { classes } from 'common/react';
+import { Component, findDOMfromVNode, InfernoNode, render } from 'inferno';
+import { Box, BoxProps } from './Box';
+import { Icon } from './Icon';
+
+export interface DropdownEntry {
```
[the remark this answers is not in the document]
Comment (actioninja) [reply tgui/packages/tgui/components/Dropdown.tsx:7]: Not really a super meaningful difference, I picked interface mostly on vibes. It more explicitly communicated to me "object that has these fields" vs "object with these specific fields."

File: tgui/packages/tgui/components/Dropdown.tsx (hunk 2 above)
Comment (actioninja) [reply tgui/packages/tgui/components/Dropdown.tsx:111]: the ts disable is for for the `domNode.offsetWidth` line, the issue is that `domNode` is a var acquired via an already very hacky method and there isn't a safe way to cast it to specify.

**Pull request merged**